### PR TITLE
refactor: remove hard coded `'Y-m-d H:i:s'`

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
See https://github.com/codeigniter4/shield/issues/608 and https://github.com/codeigniter4/CodeIgniter4/pull/8538

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: src/Models/LoginModel.php [hunk 1]
```diff
@@ -40,7 +40,7 @@ class LoginModel extends BaseModel
         'identifier' => 'permit_empty|string',
         'user_agent' => 'permit_empty|string',
         'user_id'    => 'permit_empty|integer',
-        'date'       => 'required|valid_date',
+        'date'       => 'required',
```
Comment (kenjis) [inline src/Models/LoginModel.php:43]: Do we need more validation rule?

**Comment by kenjis** [comment]
I sent PR to fix: https://github.com/codeigniter4/CodeIgniter4/pull/8543

```
 ------ --------------------------------------------------------------------- 
  Line   src/Controllers/MagicLinkController.php                              
 ------ --------------------------------------------------------------------- 
  108    Parameter #1 $row of method CodeIgniter\Model::insert() expects      
         array<int|string, float|int|string|null>|object|null, array<string,  
         CodeIgniter\I18n\Time|int|string|null> given.                        
 ------ --------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------- 
  Line   src/Models/LoginModel.php                                            
 ------ --------------------------------------------------------------------- 
  77     Parameter #1 $row of method CodeIgniter\Model::insert() expects      
         array<int|string, float|int|string|null>|object|null, array<string,  
         CodeIgniter\I18n\Time|int|string|null> given.                        
 ------ --------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------- 
  Line   src/Models/RememberModel.php                                         
 ------ --------------------------------------------------------------------- 
  58     Parameter #1 $row of method CodeIgniter\Model::insert() expects      
         array<int|string, float|int|string|null>|object|null, array<string,  
         CodeIgniter\I18n\Time|int|string|null> given.                        
 ------ --------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------- 
  Line   tests/Authentication/MagicLinkTest.php                               
 ------ --------------------------------------------------------------------- 
  113    Parameter #1 $row of method CodeIgniter\Model::insert() expects      
         array<int|string, float|int|string|null>|object|null, array<string,  
         CodeIgniter\I18n\Time|int|string|null> given.                        
  135    Parameter #1 $row of method CodeIgniter\Model::insert() expects      
         array<int|string, float|int|string|null>|object|null, array<string,  
         CodeIgniter\I18n\Time|int|string|null> given.                        
 ------ --------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------- 
  Line   tests/Controllers/MagicLinkTest.php                                  
 ------ --------------------------------------------------------------------- 
  102    Parameter #1 $row of method CodeIgniter\Model::insert() expects      
         array<int|string, float|int|string|null>|object|null, array<string,  
         CodeIgniter\I18n\Time|int|string|null> given.                        
 ------ --------------------------------------------------------------------- 

Error:  [ERROR] Found 6 errors
```
https://github.com/codeigniter4/shield/actions/runs/7895251231/job/21547330117?pr=1027

File: src/Authentication/Authenticators/Session.php [hunk 2]
```diff
@@ -922,17 +922,17 @@ protected function rememberUser(User $user): void
             $user,
             $selector,
             $this->hashValidator($validator),
-            $expires
+            $expires->format('Y-m-d H:i:s')
```
Comment (datamweb) [inline src/Authentication/Authenticators/Session.php:925]: why !? :monocle_face:

File: src/Models/LoginModel.php (hunk 1 above)
Comment (datamweb) [reply src/Models/LoginModel.php:43]: The value of **date** is not taken from the user. Therefore, I do not see the need `valid_date`.

File: src/Authentication/Authenticators/Session.php (hunk 2 above)
Comment (datamweb) [reply src/Authentication/Authenticators/Session.php:925]: Ah, I saw @TODO.

**Review by datamweb — COMMENTED** [review]
**dateFormat** has been added to CI version 4.5 In contrast, Shield supports version 4.3.5. What effect will these changes have on older versions?

File: src/Authentication/Authenticators/Session.php (hunk 2 above)
Comment (kenjis) [reply src/Authentication/Authenticators/Session.php:925]: Yes, we must pass a string. Otherwise it will be a breaking change.

**Comment by kenjis** [comment]
> **dateFormat** has been added to CI version 4.5 In contrast, Shield supports version 4.3.5. What effect will these changes have on older versions?

In my understanding, there is no changes.
The `Time` instance that is passed to a Model will be converted to a string with hard coded `Y-m-d H:i:s` in the Model.
See https://github.com/codeigniter4/CodeIgniter4/blob/3595944eac90df7e499403bad8767935c41488d9/system/BaseModel.php#L1316-L1320

File: src/Authentication/Authenticators/Session.php [hunk 3]
```diff
@@ -922,17 +922,17 @@ protected function rememberUser(User $user): void
...
         $this->setRememberMeCookie($rawToken);
     }
 
-    private function calcExpires(): string
+    private function calcExpires(): Time
     {
         $timestamp = Time::now()->getTimestamp() + setting('Auth.sessionConfig')['rememberLength'];
 
-        return Time::createFromTimestamp($timestamp)->format('Y-m-d H:i:s');
+        return Time::createFromTimestamp($timestamp);
```
Comment (datamweb) [inline src/Authentication/Authenticators/Session.php:935]: @kenjis This issue is not related to this PR, but I realized it now.
It seems that method `createFromTimestamp()` is not affected by changing the value of `App::$appTimezone` .

For example, by changing the value to App::$appTimezone = 'Asia/Tehran', I would expect the `createFromTimestamp()` method to be affected, but **timezone** is still set to 'UCT', which doesn't seem right. look :
https://github.com/codeigniter4/CodeIgniter4/blob/7d50bd33056cd54aef7f1905c1304947409ef078/system/I18n/TimeTrait.php#L266-L272

To fix it, you must do the following:
```diff
+use Config/App;
-return Time::createFromTimestamp($timestamp);
+return Time::createFromTimestamp($timestamp, config(App::class)->appTimezone);
```
Or make changes in CI4 
https://github.com/codeigniter4/CodeIgniter4/blob/7d50bd33056cd54aef7f1905c1304947409ef078/system/I18n/TimeTrait.php#L266-L272

Comment (kenjis) [reply src/Authentication/Authenticators/Session.php:935]: The current behavior seems to be misleading.
It seems better to change `Time`.

Carbon sets the current timezone. 
> You can create instances from [unix timestamps](http://en.wikipedia.org/wiki/Unix_time). createFromTimestamp() create a Carbon instance equal to the given timestamp and will set the timezone as well or default it to the current timezone. 
https://carbon.nesbot.com/docs/#api-instantiation

Comment (kenjis) [reply src/Authentication/Authenticators/Session.php:935]: I sent a PR: https://github.com/codeigniter4/CodeIgniter4/pull/8544

Comment (kenjis) [reply src/Authentication/Authenticators/Session.php:935]: Fixed by v4.4.6.

**Review by datamweb — APPROVED** [review]

**Comment by datamweb** [comment]
@kenjis thank you!

**Pull request merged**